extensions/ql-vscode/src/data-serialization/json-validator.ts

```diff
@@ -0,0 +1,50 @@
+import Ajv from "ajv";
+import { JsonSchemaTypes } from "./json-schema-types";
+import * as fs from "fs";
+import { JSONSchema7 as JSONSchema } from "json-schema";
+
+export class JsonValidator {
+  ajv: Ajv;
+  schemaTypes: string[];
+  schemas: { [key: string]: JSONSchema };
+
+  constructor(schemaTypes?: string[]) {
+    this.ajv = new Ajv();
+    this.schemaTypes = schemaTypes || JsonSchemaTypes;
+    this.schemas = this.loadSchemas(this.schemaTypes);
+  }
+
+  public validate(object: unknown, typeName: string) {
```

Comment:
I think if we try we can give this method a return type that's not `unknown` or `any`.

Take a look at https://github.com/github/codeql-variant-analysis-action/blob/main/src/json-validation.ts for inspiration. By using `SchemaTypes` as a map from keys to types, and then `Schema = keyof SchemaTypes` and `validators: Record<Schema, ValidateFunction>`, we were able to make the return type of `validateObject` be the correct object type. This means that there are no casts or objects typed as `any` outside of the validator function.

I'm also happy to help take a stab at it since it's quite arcane and complicated getting the type system to behave.

Reply:
Ah yes, I did see that. I'll write a follow-up ticket to poke at improving this since this PR is getting mahoosive. Thanks for explaining it. 👍